design/service-apis.md

```diff
@@ -0,0 +1,384 @@
+Contour release. In this case, a Gateway of this class will be automatically upgraded when the operator is upgraded.
+Initially, `projectcontour.io/contour-operator` will be the only supported value. When support is added,
+`projectcontour.io/contour-operator/<version>` will be used to version GatewayClasses. Versioning a GatewayClass will
+consist of the following workflow:
+
+- The Platform Provider upgrades the operator image from x to y, triggering an upgrade of the operator. The
+  Contour CRD is updated if y includes API changes.
+- The status of `projectcontour.io/contour-operator/x` GatewayClasses are updated to "Waiting" and associated
+  Gateways are updated to "NoSuchGatewayClass". The details for accomplishing this step is TBD (see the Open Questions
+  section).
+- The operator will no longer manage Gateways of `projectcontour.io/contour-operator/x` GatewayClasses. These
+  Gateways will continue to function and traffic will continue to be proxied. Contour/Envoy will continue running x.
+- Optionally, the Platform Provider updates the Contour resource referenced by the GatewayClass, or creates and
+  associates a new Contour to the GatewayClass. For example, a new configuration field is exposed through the Contour
+  resource in y.
+- The Platform Provider updates Gateways by setting GatewayClasses to `projectcontour.io/contour-operator/y`. This
+  triggers the operator to upgrade Contour/Envoy to y and optionally update the configuration.
+  GatewayClass|Gateway|xRoute status conditions are updated accordingly.
+
+## Implementation Details
+
+### Controller
+
+Update the existing Contour controller to only instantiate a Contour instance if `contour.spec.gatewayClassRef` is
+"None".
+
+Create a controller that reconciles the GatewayClass resource and performs the following:
+
+- Only reconcile a GatewayClass that specifies `controller: projectcontour.io/contour-operator`.
+- Ensure the Contour referenced by a GatewayClass `parametersRef` exists and is valid.
+- Update the GatewayClass status conditions accordingly.
+
+Create a controller that reconciles the Gateway resource and performs the following:
+
+- Ensure the GatewayClass specified by `gatewayClassName` exists and add the
+  `gateway-exists-finalizer.networking.x-k8s.io` finalizer to a GatewayClass whenever an associated Gateway is running.
+- Validate the `spec` configuration.
+- Ensure an instance of Contour exists with the desired configuration.  
+- Update the Gateway status conditions accordingly.
+
+### RBAC
+
+Existing Contour Operator RBAC policies must be modified to allow the `update` verb for GatewayClass and Gateway
+resources.
+
+### Test Plan
+
+- Add e2e, integration, and unit tests.
+- Create a CI job to run tests.
+
+### Open Questions
+
+- For versioned GatewayClass support, i.e. `controller: projectcontour.io/contour-operator/v1.12.0`, should Contour be
```

Comment:
I think we're better to punt more careful upgrades down the road a bit and say that when you upgrade the operator to a version, it will immediately upgrade all controllers under its management. That way we can work on getting the intial install working first, which is after all a prerequisite for being able to upgrade.